query-parser/src/parse.rs

```diff
@@ -83,320 +96,348 @@ error_chain! {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "find"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_with, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "with"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_where, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "where"
+        } else {
+            false
+        }
+    })
+});
+
+/// Express something close to a builder pattern for a `FindQuery`.
+enum FindQueryPart {
+    FindSpec(FindSpec),
+    With(Vec<Variable>),
+    WhereClauses(Vec<WhereClause>),
 }
 
-#[allow(dead_code)]
-pub fn clause_seq_to_patterns(clauses: &[edn::Value]) -> WhereParseResult {
-    Where::clauses()
-        .parse(clauses)
+/// This is awkward, but will do for now.  We use `keyword_map()` to optionally accept vector find
+/// queries, then we use `FindQueryPart` to collect parts that have heterogeneous types; and then we
+/// construct a `FindQuery` from them.
+def_parser!(Find, query, FindQuery, {
+    let p_find_spec = Find::literal_find()
+        .with(vector().of(Find::spec().map(FindQueryPart::FindSpec)));
+
+    let p_with_vars = Find::literal_with()
+        .with(vector().of(many(Query::variable()).map(FindQueryPart::With)));
+
+    let p_where_clauses = Find::literal_where()
+        .with(vector().of(Where::clauses().map(FindQueryPart::WhereClauses))).expected(":where clauses");
+
+    (or(map(), keyword_map()))
+        .of(many(choice::<[&mut Parser<Input = ValueStream, Output = FindQueryPart>; 3], _>([
+            &mut try(p_find_spec),
+            &mut try(p_with_vars),
+            &mut try(p_where_clauses),
+        ])))
+        .and_then(|parts: Vec<FindQueryPart>| -> std::result::Result<FindQuery, combine::primitives::Error<edn::ValueAndSpan, edn::ValueAndSpan>>  {
+            let mut find_spec = None;
+            let mut with_vars = None;
+            let mut where_clauses = None;
+
+            for part in parts {
+                match part {
+                    FindQueryPart::FindSpec(x) => find_spec = Some(x),
+                    FindQueryPart::With(x) => with_vars = Some(x),
+                    FindQueryPart::WhereClauses(x) => where_clauses = Some(x),
+                }
+            }
+
+            Ok(FindQuery {
+                find_spec: find_spec.clone().ok_or(combine::primitives::Error::Unexpected("expected :find".into()))?,
+                default_source: SrcVar::DefaultSrc,
+                with: with_vars.unwrap_or(vec![]),
+                in_vars: vec![],       // TODO
+                in_sources: vec![],    // TODO
+                where_clauses: where_clauses.ok_or(combine::primitives::Error::Unexpected("expected :where".into()))?,
```

Comment:
Wait a minute, doesn't this discard any parse error that comes out of parsing `p_where_clauses`, defeating the whole point of stacking this whole thing into parser-land?